sklearn/neighbors/_nearest_centroid.py

```diff
@@ -162,73 +169,86 @@ def fit(self, X, y):
+
         check_classification_targets(y)
 
-        n_samples, n_features = X.shape
         le = LabelEncoder()
         y_ind = le.fit_transform(y)
         self.classes_ = classes = le.classes_
+
         n_classes = classes.size
         if n_classes < 2:
             raise ValueError(
-                "The number of classes has to be greater than one; got %d class"
-                % (n_classes)
+                "The number of classes has to be greater than one; got"
+                f" {n_classes} class."
             )
 
-        # Mask mapping each class to its members.
-        self.centroids_ = np.empty((n_classes, n_features), dtype=np.float64)
-        # Number of clusters in each class.
-        nk = np.zeros(n_classes)
-
-        for cur_class in range(n_classes):
-            center_mask = y_ind == cur_class
-            nk[cur_class] = np.sum(center_mask)
+        # Choose the function from computing centroids.
+        if self.metric == "manhattan":
+            # NumPy does not calculate median of sparse matrices.
             if is_X_sparse:
-                center_mask = np.where(center_mask)[0]
-
-            if self.metric == "manhattan":
-                # NumPy does not calculate median of sparse matrices.
-                if not is_X_sparse:
-                    self.centroids_[cur_class] = np.median(X[center_mask], axis=0)
-                else:
-                    self.centroids_[cur_class] = csc_median_axis_0(X[center_mask])
+                compute_centroid = csc_median_axis_0
             else:
-                # TODO(1.5) remove warning when metric is only manhattan or euclidean
-                if self.metric != "euclidean":
-                    warnings.warn(
-                        "Averaging for metrics other than "
-                        "euclidean and manhattan not supported. "
-                        "The average is set to be the mean."
-                    )
-                self.centroids_[cur_class] = X[center_mask].mean(axis=0)
-
-        if self.shrink_threshold:
+                compute_centroid = partial(np.median, axis=0)
+        else:
+            # TODO(1.5) remove warning when metric is only manhattan or euclidean
+            if self.metric != "euclidean":
+                warnings.warn(
+                    "Averaging for metrics other than "
+                    "euclidean and manhattan not supported. "
+                    "The average is set to be the mean."
+                )
+            compute_centroid = partial(np.mean, axis=0)
+
+        # Choose the transformation for boolean class mask vector.
+        if is_X_sparse:
+            mask_trf = lambda mask: np.where(mask)[0]  # noqa: E731
+        else:
+            mask_trf = lambda mask: mask  # noqa: E731
+
+        n_samples, n_features = X.shape
+
+        # Compute the centroids.
+        self.centroids_ = np.empty((n_classes, n_features), dtype=np.float64)
```

Comment:
Actually, I am wondering if this refactoring is necessary since we are modifying this code in this PR: https://github.com/scikit-learn/scikit-learn/pull/26689

I have to check if we can just combine both approaches.